File: src/towd/Persister.vb
Imports System.IO
Imports System.Text.Json
Imports towd.data
Imports towd.ui

Friend Class Persister
    Implements IPersister

    Public Function SaveGame(SaveSlot As ISaveSlot, worldData As WorldData) As Task Implements IPersister.SaveGame
        File.WriteAllText(SaveSlot.Filename, JsonSerializer.Serialize(worldData))
        Return Task.CompletedTask
    End Function

    Public Function SaveExists(saveSlot As ISaveSlot) As Task(Of DateTime?) Implements IPersister.SaveExists
        If File.Exists(saveSlot.Filename) Then
            Return Task.FromResult(Of Date?)(File.GetLastWriteTime(saveSlot.Filename))
        Else
            Return Task.FromResult(Of Date?)(Nothing)
        End If
    End Function

    Public Function LoadGame(saveSlot As ISaveSlot) As Task(Of WorldData) Implements IPersister.LoadGame
        Try
            Return Task.FromResult(JsonSerializer.Deserialize(Of WorldData)(File.ReadAllText(saveSlot.Filename)))
        Catch
            Return Task.FromResult(Of WorldData)(Nothing)
        End Try
    End Function
End Class

File: src/towd.blazor.standalone/Persister.cs
using Microsoft.JSInterop;
using System.Text.Json;
using towd.data;
using towd.ui;

namespace towd.blazor.standalone
{
    public class Persister(IJSRuntime JSRuntime) : IPersister
    {
        public async Task<WorldData> LoadGame(ISaveSlot saveSlot)
        {
            var data = await JSRuntime.InvokeAsync<string>("loadGame", saveSlot.Filename);
            if(data != null)
            {
#pragma warning disable CS8603 // Possible null reference return.
                return JsonSerializer.Deserialize<WorldData>(data);
#pragma warning restore CS8603 // Possible null reference return.
            }
#pragma warning disable CS8603 // Possible null reference return.
            return null;
#pragma warning restore CS8603 // Possible null reference return.
        }

        public async Task<DateTime?> SaveExists(ISaveSlot saveSlot)
        {
            var result = await JSRuntime.InvokeAsync<string>("saveExists", saveSlot.Filename);
            if(result==null)
            {
                return null;
            }
            return DateTime.Parse(result);
        }

        public async Task SaveGame(ISaveSlot saveSlot, WorldData worldData)
        {
            await JSRuntime.InvokeVoidAsync("saveGame", saveSlot.Filename, JsonSerializer.Serialize(worldData));
        }
    }
}
